pkg/yurtctl/cmd/convert/convert.go

```diff
@@ -303,11 +311,19 @@ func (co *ConvertOptions) RunConvert() (err error) {
 	for _, node := range nodeLst.Items {
 		if !strutil.IsInStringLst(co.CloudNodes, node.GetName()) || strutil.IsInStringLst(kcmNodeNames, node.GetName()) {
 			if !isNodeReady(&node.Status) {
-				klog.Errorf("Cannot do the convert, the status of worker node or kube-controller-manager node: %s is not 'Ready'.", node.Name)
+				err = fmt.Errorf("the status of worker node or kube-controller-manager node: %s is not 'Ready'", node.Name)
 				return
 			}
 		}
 	}
+
+	// 1.3 check the nodes label
```

Comment:
you see, we label the node before we actually do the converting thing.
so, converting failed may lead to the node being labeled but users have to `yurtctl convert` again
because the log tells them failed.

so it leads to chooses:
1. keep it. Users must know, they have to run `yurt revert`  after failing when they want to do it again. And they may not know and often forget.
2. we make `yurtctl convert` idempotent which means I don't care if nodes have been converted or not, I can do it whenever I want, with no mind pressure or side effect.
3. we make sure only if the node has been converted successfully we label the node. so we have to unlabeled the node after failure or move labeling to the end.

personal I like the way 2, because it brings a good experience. but it introduces complicated work too.

Reply:
way 2 <-- agree +1

Reply:
@adamzhoul @rambohe-ch I agree with you that way 2 is the most user-friendly way, but it also increases the complexity of `yurtctl convert`.  If we make `yurtctl convert` idempotent, we have to consider every change made by `yurtctl convert` and figure an idempotent way and sometimes lead to tricky situations.

For example, the user has converted the cluster and was enabling autonomy on node1 and node2. Then the user finds that only node2 should enable autonomy feature,  so in the next converting, the user only configures node1 for node autonomy. In this case, should we cancel the node autonomy for node2?

IMO, as we are going to introduce the precheck command for `yurtctl` and I think it is enough to prevent users from `yurtctl convert` failure. 

Besides, `kubeadm init/join` is not idempotent. I think as long as we can provide an acceptable success rate for `yurtctl convert`, then idempotent is not very necessary.

Reply:
## What do I think of idempotent?

I don't know why I failed last time, I don't know why and don't know what to do.
try again is the most natural reaction and the only way I know.

and it is something like `ultimate state`, when I typed it I want it.

##  if implemented idempotent

> For example, the user has converted the cluster and was enabling autonomy on node1 and node2. Then the user finds that only node2 should enable autonomy feature, so in the next converting, the user only configures node1 for node autonomy. In this case, should we cancel the node autonomy for node2?

I think the command: `yurtctl convert ...`  represents the ultimate state user expect.
so,  we don't care what the current state is, failed job、 different label, etc...
just, leads the cluster to what users expect.

as for this case: yes, we should cancel node2

> IMO, as we are going to introduce the precheck command for yurtctl and I think it is enough to prevent users from yurtctl convert failure.

some failing may not be checked by precheck. like, pull image timeout.
I think we have  got much feedback from the community 
something like:
```
E1125 21:09:13.772921   19268 util.go:539] fail to run servant job(yurtctl-servant-convert-k8smaster): jobs.batch "yurtctl-servant-convert-k8smaster" already exists
```
they do know `yurtctl convert` failed, that's all.
they don't know what the job log means and don't know what to do.



## Do we have to implement it this time?

not necessary I think. 
it may lead to too many updates, we should think more carefully.

> I think as long as we can provide an acceptable success rate for yurtctl convert, then idempotent is not very necessary.

if we want to handle failure issues, it's unnecessary if we achieved an acceptable success rate. like you said.
if we consider it is something like `ultimate state`, users can execute many many times for updates. it's gonna be a capability we give to users.

## What should we fix this time I think?

the label or failed job should not stop users from trying again.
unless we are sure the label can really represent the job is done well.

Reply:
@adamzhoul Hi, according to our previous discussion, we will not cover idempotent in this PR, so I leave it unchanged.